ramls/loginResponseWithExpiry.json

```diff
@@ -0,0 +1,19 @@
+{
+  "$schema": "http://json-schema.org/draft-04/schema#",
+  "description": "Object returned on successful login or on token refresh",
+  "type": "object",
+  "properties": {
+    "accessTokenExpiration": {
+      "description": "The time in UTC after which the access token will be considered expired. Time format is ISO 8601.",
+      "type": "string"
+    },
+    "refreshTokenExpiration": {
+      "description": "The time in UTC after which the refresh token will be considered expired. Time format is ISO 8601.",
+      "type": "string"
+    }
```

Comment:
I think the best way to think of `/login-with-expiry` is as login for the least secure client (the browser context) and probably nothing else at least for now. As part of our original proposal we had the concept of an API Token -- an unencrypted non-expiring token with non-user based permissions that sys admins could manage and use in more secure systems like edge modules. Some of the work has been done to make this happen in mod-authtoken, but not all. Instead of finishing the API token stuff first, the decision was made to move forward with the stuff that will have the most impact on security first and then circle back to API tokens.

In other words we're not getting rid of the old `/login` route, but we are shutting it down for browser clients by doing the work to have those clients now use `/login-with-expiry`. So edge-apis and others (like developers!) can still get a non expiring "legacy" token from the old `/login` endpoint for the time being.

Once we have something like API tokens we'll be in a better position to fully depreciate `/login`.

Reply:
> I think the best way to think of `/login-with-expiry` is as login for the least secure client (the browser context) and probably nothing else at least for now. As part of our original proposal we had the concept of an API Token -- an unencrypted non-expiring token with non-user based permissions that sys admins could manage and use in more secure systems like edge modules. Some of the work has been done to make this happen in mod-authtoken, but not all. Instead of finishing the API token stuff first, the decision was made to move forward with the stuff that will have the most impact on security first and then circle back to API tokens.

Thank you for explaining this to me. 

My understanding was that all clients would be moving over to use refresh tokens and the regular non-expiring tokens would only be retained for the transition. It now seems that the login endpoint will kept for longer until we have a suitable solution for non-browser based clients.

I think it could be useful to provide some documentation to explain these design decisions and provide guidance to folks who need to choose which endpoints to use.

> In other words we're not getting rid of the old `/login` route, but we are shutting it down for browser clients by doing the work to have those clients now use `/login-with-expiry`. So edge-apis and others (like developers!) can still get a non expiring "legacy" token from the old `/login` endpoint for the time being.

How is the login endpoint being shut down for browser clients?

> Once we have something like API tokens we'll be in a better position to fully depreciate `/login`.

Ok.

Reply:
Even in the stripes based browser UI case, I think that including the expiration information in the body of the response creates a reliance on both the headers (cookies) and the body of the message, which could have been avoided.

The expiration information could have been included in separate non HTTP-only cookies (or other headers) instead of changing the body.

Reply:
If I'm understanding, you're saying that it seems wrong here to provide information about token expiration in the body of the response, maybe especially because that body is semantically defined as a `CompositeUser` (not here but in mod-users-bl).

I think my impulse was to provide the expiration in a convenient location and hadn't really considered the semantic issue perhaps that you're raising but about mod-users-bl.

I think another way of making the argument might be: use headers for auth stuff consistently across modules since some modules (like mod-users-bl) might want to return other stuff in the body. I could per persuaded about that.

Reply:
I don't see any problem with the expiration information in the body of POST `/login-with-expiry`.

> The expiration information could have been included in separate non HTTP-only cookies

No additional cookie should be added. https://datatracker.ietf.org/doc/html/rfc6265#page-27 says:

> Servers SHOULD use as few and as small cookies as possible

If there is a need a GET `/token/expiration` API can be added that returns the access and refresh token exiration in the body.